File: src/main/java/at/Xirado/htl/bot/listeners/ButtonListener.java
package at.Xirado.htl.bot.listeners;

import at.Xirado.htl.bot.Main;
import at.Xirado.htl.bot.misc.Util;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.events.interaction.ButtonClickEvent;
import net.dv8tion.jda.api.exceptions.ErrorHandler;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import net.dv8tion.jda.api.interactions.components.Button;
import net.dv8tion.jda.api.requests.ErrorResponse;
import org.jetbrains.annotations.NotNull;

import java.awt.*;

import static at.Xirado.htl.bot.listeners.OnGuildMemberJoin.*;

public class ButtonListener extends ListenerAdapter
{
    public static final Button JOIN_SERVER = Button.link(Util.INVITE_URL, "HTL-Discord beitreten");

    @Override
    public void onButtonClick(@NotNull ButtonClickEvent event)
    {
        Role role = event.getJDA().getRolesByName("Verifiziert", false).get(0);
        switch (event.getInteraction().getComponentId())
        {
            case "send_again" -> {
                Member member = event.getMember();
                if (member == null) return;
                if (member.getRoles().contains(role))
                {
                    MessageEmbed embed = new EmbedBuilder()
                            .setDescription("Du bist bereits verifiziert!")
                            .setAuthor("Verifizierung | HTL Discord", null, Util.LOGO_URL)
                            .setColor(0xCF1020)
                            .build();
                    event.replyEmbeds(embed).setEphemeral(true).queue();
                    return;
                }
                User user = event.getUser();
                event.deferReply(true)
                        .flatMap(x -> user.openPrivateChannel())
                        .flatMap(channel -> channel.sendMessageEmbeds(DM_EMBED).setActionRow(VERIFY_BUTTON, RULES_BUTTON))
                        .mapToResult()
                        .flatMap(result -> {
                            if (result.isFailure())
                            {
                                EmbedBuilder builder = new EmbedBuilder()
                                        .setColor(0xCF1020)
                                        .setDescription("Ich konnte dir keine Nachricht schicken, da du Privatnachrichten deaktiviert hast!\nWenn du dir sicher bist, dass es nicht an dir liegt, sende bitte eine DM an <@184654964122058752>")
                                        .setAuthor("Verifizierung | HTL Discord", null, Util.LOGO_URL);
                                return event.getHook().sendMessageEmbeds(builder.build());
                            } else
                            {
                                EmbedBuilder builder = new EmbedBuilder()
                                        .setColor(0x3EB489)
                                        .setAuthor("Verifizierung | HTL Discord", null, Util.LOGO_URL)
                                        .setDescription("Du hast eine Nachricht von mir erhalten!");
                                return event.getHook().sendMessageEmbeds(builder.build());
                            }
                        }).queue();
            }
            case "verify" -> {
                if (role != null)
                {
                    event.deferReply().queue(x -> {
                        Guild guild = role.getGuild();
                        guild.retrieveMember(event.getUser()).queue(member -> {
                            if (member.getRoles().contains(role))
                            {
                                MessageEmbed embed = new EmbedBuilder()
                                        .setDescription("Du bist bereits verifiziert!")
                                        .setAuthor("Verifizierung | HTL Discord", null, Util.LOGO_URL)
                                        .setColor(0xCF1020)
                                        .build();
                                x.sendMessageEmbeds(embed).queue();
                                return;
                            }
                            guild.addRoleToMember(member, role)
                                    .flatMap(y -> {
                                        MessageEmbed embed = new EmbedBuilder()
                                                .setDescription("Danke, du wurdest erfolgreich verifiziert. **Wir wünschen dir viel Spaß!**")
                                                .setColor(0x3EB489)
                                                .setAuthor("Verifizierung | HTL Discord", null, Util.LOGO_URL)
                                                .build();
                                        return x.sendMessageEmbeds(embed);
                                    }).queue();

                        }, new ErrorHandler().handle(ErrorResponse.UNKNOWN_MEMBER, (e) -> x.sendMessage("Du bist nicht auf dem Server!").queue()));
                    });
                }
            }
        }
    }
}

File: src/main/kotlin/at/xirado/htl/listener/ButtonListener.kt
package at.xirado.htl.listener

import at.xirado.htl.*
import dev.minn.jda.ktx.CoroutineEventListener
import dev.minn.jda.ktx.await
import net.dv8tion.jda.api.events.GenericEvent
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent
import net.dv8tion.jda.api.exceptions.ErrorResponseException
import net.dv8tion.jda.api.requests.ErrorResponse

class ButtonListener : CoroutineEventListener {

    override suspend fun onEvent(event: GenericEvent) {
        when (event) {
            is ButtonInteractionEvent -> onButtonInteraction(event)
        }
    }

    private suspend fun onButtonInteraction(event: ButtonInteractionEvent) {
        val role = event.jda.getRoleById(VERIFIED_ROLE_ID)
        when (event.interaction.componentId) {
            "send_again" -> {
                if (role == null) {
                    event.replyEmbeds(ERROR_OCCURRED_EMBED).setEphemeral(true).queue()
                    log.error("Verified Role no longer exists!")
                    return
                }
                val member = event.member
                if (role in member!!.roles) {
                    event.replyEmbeds(ALREADY_VERIFIED_EMBED).setEphemeral(true).queue()
                    return
                }

                val user = member.user
                event.deferReply(true).queue()
                try {
                    val channel = user.openPrivateChannel().await()
                    channel.sendMessageEmbeds(DM_EMBED).setActionRow(VERIFY_BUTTON, RULES_BUTTON).await()
                    event.hook.sendMessageEmbeds(DM_SENT_SUCCESS_EMBED).await()
                } catch (ex: ErrorResponseException) {
                    if (ex.errorResponse == ErrorResponse.CANNOT_SEND_TO_USER) {
                        event.hook.sendMessageEmbeds(DM_DISABLED_EMBED).await()
                        return
                    }
                    log.error("An unexpected error occurred! (User: ${user.idLong} / ${user.asTag})", ex)
                    event.hook.sendMessageEmbeds(ERROR_OCCURRED_EMBED).await()
                }
            }
            "verify" -> {
                if (role == null) {
                    event.replyEmbeds(ERROR_OCCURRED_EMBED).queue()
                    log.error("Verified Role no longer exists!")
                    return
                }
                event.deferReply().queue()
                val guild = role.guild
                try {
                    val member = guild.retrieveMember(event.user).await()
                    if (member.hasRole(VERIFIED_ROLE_ID)) {
                        event.hook.sendMessageEmbeds(ALREADY_VERIFIED_EMBED).queue()
                        return
                    }
                } catch (ex: ErrorResponseException) {
                    if (ex.errorResponse == ErrorResponse.UNKNOWN_MEMBER) {
                        event.hook.sendMessageEmbeds(NO_LONGER_IN_GUILD_EMBED).addActionRow(INVITE_BUTTON).await()
                        return
                    } else {
                        event.hook.sendMessageEmbeds(ERROR_OCCURRED_EMBED).await()
                        log.error("An unexpected error occurred!", ex)
                        return
                    }
                }
                try {
                    guild.addRoleToMember(event.user.idLong, role).await()
                    event.hook.sendMessageEmbeds(VERIFIED_SUCCESS_EMBED).await()
                } catch (ex: Exception) {
                    event.hook.sendMessageEmbeds(ERROR_OCCURRED_EMBED).await()
                    log.error("An unexpected error occurred!", ex)
                }
            }
        }
    }
}